includes/subscriptions/class-wc-payments-subscriptions-event-handler.php

```diff
@@ -0,0 +1,190 @@
+	 * @param WC_Payments_Subscription_Service $subscription_service Subscription service.
+	 */
+	public function __construct( WC_Payments_Invoice_Service $invoice_service, WC_Payments_Subscription_Service $subscription_service ) {
+		$this->invoice_service      = $invoice_service;
+		$this->subscription_service = $subscription_service;
+	}
+
+	/**
+	 * Adds fee, discount, and shipping related invoice items to WCPay subscription.
+	 *
+	 * @param array $body The event body that triggered the webhook.
+	 *
+	 * @throws Rest_Request_Exception Required parameters not found.
+	 */
+	public function handle_invoice_upcoming( array $body ) {
+		$event_data            = $this->get_event_property( $body, 'data' );
+		$event_object          = $this->get_event_property( $event_data, 'object' );
+		$wcpay_subscription_id = $this->get_event_property( $event_object, 'subscription' );
+		$wcpay_customer_id     = $this->get_event_property( $event_object, 'customer' );
+		$subscription          = WC_Payments_Subscription_Service::get_subscription_from_wcpay_subscription_id( $wcpay_subscription_id );
+
+		if ( ! $subscription ) {
+			throw new Rest_Request_Exception( __( 'Cannot find subscription to handle the "invoice.upcoming" event.', 'woocommerce-payments' ) );
+		}
+
+		$wcpay_subscription = $this->subscription_service->get_wcpay_subscription( $subscription );
+
+		// Suspend or cancel subscription if we didn't expect a next payment.
+		if ( 0 === $subscription->get_time( 'next_payment' ) ) {
+			// TODO: Add error handling to these {cancel/suspend}_subscription calls i.e. add a subscription order note if the WCPay subscription wasn't cancelled.
+			if ( ! $subscription->has_status( 'on-hold' ) && 0 !== $subscription->get_time( 'end' ) ) {
+				$this->subscription_service->cancel_subscription( $subscription );
+				$subscription->add_order_note( __( 'There was an upcoming payment event however the subscription is due to end in WooCommerce. The subscription has been cancelled.', 'woocommerce-payments' ) );
+			} else {
+				$this->subscription_service->suspend_subscription( $subscription );
+				$subscription->add_order_note( __( 'There was an upcoming payment event however the subscription is on-hold. The subscription has been suspended.', 'woocommerce-payments' ) );
+			}
+		} else {
+			// Update the subscription in WC to match the WCPay Subscription's next payment date.
+			$this->subscription_service->update_dates_to_match_wcpay_subscription( $wcpay_subscription, $subscription );
+			// Translators: %s Scheduled/upcoming payment date in Y-m-d H:i:s format.
+			$subscription->add_order_note( sprintf( __( "There's an upcoming invoice which will automatically attempt payment on %s. The subscription's next payment date has been updated to match.", 'woocommerce-payments' ), get_date_from_gmt( gmdate( 'Y-m-d H:i:s', $wcpay_subscription['current_period_end'] ), wc_date_format() . ' ' . wc_time_format() ) ) );
```

Comment:
Yeah I'm not actually a big fan of this order note because if the subscription dates weren't updated, we're still adding this note to say they were.

Also, if the subscription next payment date was only updated by a few seconds/minutes then an order note feels a little unnecessary. On the other hand, if there was a huge discrepancy between the dates we had in Woo and the next scheduled invoice payment date then I do think it's nice to explain why the subscription's next payment was moved to x date. Maybe this sort of thing will come up in user testing.

In the latest changes I've move this note to be inside `update_dates_to_match_wcpay_subscription` and updated message to be trimmed down a bit.

```
Next automatic payment scheduled for %s. Subscription dates have been updated to match.
```

We could be even more concise/vague and just say:

```
Subscription dates updated to match WCPay Server.
```

But we can iterate on this

Reply:
> Also, if the subscription next payment date was only updated by a few seconds/minutes then an order note feels a little unnecessary. On the other hand, if there was a huge discrepancy between the dates we had in Woo and the next scheduled invoice payment date then I do think it's nice to explain why the subscription's next payment was moved to x date.

I agree.

Ideally we could make these into two seperate notes entirely. So on `invoice.upcoming` provide a general note like what you've done with:

```
Next automatic payment scheduled for %s.
```

And then in `update_dates_to_match_wcpay_subscription` we could handle a note specific to date changes. And for this particular note, it would be great if it could only appear when the date change is larger than a few hours, a day, or some other arbitrary value.

What do you think?

Reply:
I've gone with this approach and I've added an order note if the difference in time is 12 hours of more. See 3ab8bbe.

Let me know what you think :)

Reply:
This change looks good to me!